src/wallet/rpcwallet.cpp

```diff
@@ -3423,6 +3423,70 @@ UniValue z_getaddressforaccount(const UniValue& params, bool fHelp)
     return result;
 }
 
+UniValue z_listaccounts(const UniValue& params, bool fHelp)
+{
+    if (!EnsureWalletIsAvailable(fHelp))
+        return NullUniValue;
+
+    if (fHelp || params.size() > 1)
+        throw runtime_error(
+            "z_listaccounts\n"
+            "\nReturns the list of accounts created through z_getnewaccount.\n"
+            "\nResult:\n"
+            "[\n"
+            "   {\n"
+            "     \"account\": \"uint\",           (uint) The account id\n"
+            "        {\n"
+            "           \"diversifier\":  \"uint\",        (string) A diversifier used in the account\n"
+            "           \"ua\":  \"address\",              (string) The unified address corresponding to the diversifier.\n"
+            "        }\n"
+            "   }\n"
+            "]\n"
+            "\nExamples:\n"
+            + HelpExampleCli("z_listaccounts", "")
+        );
+
+    LOCK2(cs_main, pwalletMain->cs_wallet);
+
+    KeyIO keyIO(Params());
+    UniValue ret(UniValue::VARR);
+
+    auto hdChain = pwalletMain->GetMnemonicHDChain();
+    auto num_accounts = hdChain->GetAccountCounter();
+
+    for (int acctId = 0; acctId < num_accounts; acctId++) {
```

Comment:
Why iterate over the account counter rather than just iterate over `mapUnifiedAccountKeys` directly?